File: tdf-hq/src/TDF/API/Sessions.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}

module TDF.API.Sessions where

import           Data.Text (Text)
import qualified Data.ByteString.Lazy as BL
import           Servant

import           TDF.API.Types

type SessionsAPI =
  "sessions" :>
    (    QueryParam "page" Int
      :> QueryParam "pageSize" Int
      :> Get '[JSON] (Page SessionDTO)
  :<|> ReqBody '[JSON] SessionCreate :> PostCreated '[JSON] SessionDTO
  :<|> "options" :> Get '[JSON] SessionOptionsDTO
  :<|> Capture "id" Text :> Get '[JSON] SessionDTO
  :<|> Capture "id" Text :> ReqBody '[JSON] SessionUpdate :> Patch '[JSON] SessionDTO
  :<|> Capture "id" Text :> "input-list" :> Get '[JSON] [SessionInputRow]
  :<|> Capture "id" Text :> "input-list.pdf"
        :> Get '[OctetStream] (Headers '[Header "Content-Disposition" Text] BL.ByteString)
    )

File: archives/tdf-root-ui-legacy/api/sessions.ts
import { get, post, patch } from './client';
import type { Page, SessionCreate, SessionDTO, SessionOptionsDTO, SessionUpdate } from './types';

export const Sessions = {
  list: (params?: { page?: number; pageSize?: number }) => {
    const search = new URLSearchParams();
    if (params?.page) search.set('page', String(params.page));
    if (params?.pageSize) search.set('pageSize', String(params.pageSize));
    const suffix = search.toString() ? `?${search.toString()}` : '';
    return get<Page<SessionDTO>>(`/sessions${suffix}`);
  },
  create: (body: SessionCreate) => post<SessionDTO>('/sessions', body),
  detail: (id: string) => get<SessionDTO>(`/sessions/${id}`),
  update: (id: string, body: SessionUpdate) => patch<SessionDTO>(`/sessions/${id}`, body),
  options: () => get<SessionOptionsDTO>('/sessions/options'),
};
